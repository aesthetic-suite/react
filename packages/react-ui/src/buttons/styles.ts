import { ColorShade, PALETTE_TYPES, Rule, RuleMap, StateType, styles } from '@aesthetic/react';

export const styleSheet = styles((css) => {
	const paletteVariants: RuleMap = {};
	const emptyBackgroundHues: Record<Exclude<StateType, 'disabled'>, ColorShade> = {
		focused: '00',
		hovered: '00',
		selected: '10',
	};

	function composeRule(stateFactory: (state: StateType | 'base') => Rule) {
		return {
			...stateFactory('base'),
			'@selectors': {
				':focus': stateFactory('focused'),
				':hover': stateFactory('hovered'),
				':active': stateFactory('selected'),
				':disabled': stateFactory('disabled'),
			},
		};
	}

	PALETTE_TYPES.forEach((palette) => {
		// Solid
		paletteVariants[`palette:${palette} + fill:solid`] = composeRule((state) => ({
			color: css.var(`palette-${palette}-fg-${state}`),
			backgroundColor: css.var(`palette-${palette}-bg-${state}`),
			borderColor: css.var(`palette-${palette}-bg-${state}`),
		}));

		// Hollow
		paletteVariants[`palette:${palette} + fill:hollow`] = composeRule((state) => {
			const backgroundColor =
				state === 'base' || state === 'disabled'
					? 'transparent'
					: css.var(`palette-${palette}-color-${emptyBackgroundHues[state]}`);

			return {
				color: css.var(`palette-${palette}-bg-${state}`),
				backgroundColor,
				borderColor: css.var(`palette-${palette}-bg-${state}`),
			};
		});

		// Empty
		paletteVariants[`palette:${palette} + fill:empty`] = composeRule((state) => {
			const color =
				state === 'base' || state === 'disabled'
					? 'transparent'
					: css.var(`palette-${palette}-color-${emptyBackgroundHues[state]}`);

			return {
				color: css.var(`palette-${palette}-bg-${state}`),
				backgroundColor: color,
				borderColor: color,
			};
		});
	});

	return {
		button: {
			display: 'inline-flex',
			flexWrap: 'nowrap',
			alignItems: 'center',
			justifyContent: 'center',
			position: 'relative',
			verticalAlign: 'middle',
			textAlign: 'center',
			textDecoration: 'none',
			fontWeight: 'bold',
			fontFamily: css.var('typography-font-text'),
			cursor: 'pointer',
			borderColor: 'transparent',
			borderStyle: 'solid',
			borderWidth: 0,
			margin: 0,
			padding: 0,

			'@selectors': {
				// Removes weird bonus padding from button in Firefox
				'::-moz-focus-inner': {
					border: 0,
					padding: 0,
					margin: 0,
				},
			},

			'@variants': {
				...paletteVariants,
			},
		},

		buttonFluid: {
			width: '100%',
			maxWidth: '100%',
		},
	};
});
