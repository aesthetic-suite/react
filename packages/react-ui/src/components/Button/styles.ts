import { createComponentStyles } from '@aesthetic/react';

export default createComponentStyles((css) => ({
  button: css.mixin('text-df', {
    display: 'inline-flex',
    textAlign: 'center',
    minWidth: css.unit(8),
    padding: {
      leftRight: css.var('spacing-md'),
      topBottom: css.var('spacing-df'),
    },

    '@selectors': {
      // Removes weird bonus padding from button in Firefox
      '::-moz-focus-inner': {
        border: 0,
        padding: 0,
        margin: 0,
      },
    },
  }),

  button_block: {
    display: 'flex',
    width: '100%',
    whiteSpace: 'normal',
    overflow: 'hidden',
  },

  // SIZES

  button_small: css.mixin('text-sm', {
    minWidth: css.unit(6),
    padding: {
      leftRight: css.var('spacing-df'),
      topBottom: css.var('spacing-sm'),
    },
  }),

  button_large: css.mixin('text-lg', {
    minWidth: css.unit(10),
    padding: {
      leftRight: css.var('spacing-xl'),
      topBottom: css.var('spacing-lg'),
    },
  }),

  // PALETTES
}));
