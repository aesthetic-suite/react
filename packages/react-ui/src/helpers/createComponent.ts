/* eslint-disable @typescript-eslint/no-explicit-any */

import { forwardRef } from 'react';
import { CommonHtmlProps, HtmlElementType, OmitUnwantedHtmlProps } from '../types';

export function createComponent<Props extends object, Element extends HTMLElement = HTMLElement>(
	component: React.ForwardRefRenderFunction<Element, Props>,
) {
	return forwardRef(component);
}

export type DynamicProps<
	As extends React.ElementType,
	Props extends object
> = OmitUnwantedHtmlProps<React.ComponentPropsWithRef<As>> &
	Props & {
		/**
		 * Native HTML element or custom component to render as.
		 */
		as?: As;
	};

// This is based on the `React.ExoticComponent` but supports the dynamic `as` prop
// through a generic on the render function. We want the props on the consumer to change
// based on the `as` prop, but we *do not* want to type the props/ref in the implementation,
// as it makes the types too complex and we spread them into the DOM regardless.
export interface DynamicComponent<Props extends object, As extends React.ElementType> {
	<T extends As>(props: DynamicProps<T, Props>): React.ReactElement | null;
	readonly $$typeof: symbol;
	displayName?: string;
}

// We only want to return generic attributes when an HTML element,
// otherwise return the props from the custom component.
export type InferAsProps<T> = T extends HtmlElementType
	? CommonHtmlProps
	: T extends React.JSXElementConstructor<infer P>
	? P
	: {};

export function createDynamicComponent<
	Props extends object,
	As extends React.ElementType = HtmlElementType
>(component: React.ForwardRefRenderFunction<any, InferAsProps<As> & Props & { as?: As }>) {
	return (createComponent(component) as unknown) as DynamicComponent<Props, As>;
}
