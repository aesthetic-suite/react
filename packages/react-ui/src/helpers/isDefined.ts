export function isDefined(...values: unknown[]): boolean {
	return values.some((value) => typeof value !== 'undefined' && value !== null && value !== false);
}
