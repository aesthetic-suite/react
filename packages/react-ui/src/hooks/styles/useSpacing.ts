import { useMemo } from 'react';
import { SPACING_SIZES, SpacingSize, style, useStyles, useTheme } from '@aesthetic/react';
import { mapObject } from '../../helpers/mapObject';
import { mapVariants } from '../../helpers/mapVariants';
import { UtilityStyles } from '../../types';

type InlineSpacingSize = SpacingSize | 'inline';

export type Space = SpacingSize | number;

export interface SpacingProps {
	/** Apply padding to all edges. */
	spacing?: Space;
	/** Apply padding to the bottom edge. */
	spacingBottom?: Space;
	/** Apply padding to the end/trailing edge. */
	spacingEnd?: Space;
	/** Apply padding to the side edges. */
	spacingHorizontal?: Space;
	/** Apply padding to the start/leading edge. */
	spacingStart?: Space;
	/** Apply padding to the top edge. */
	spacingTop?: Space;
	/** Apply padding to the top and bottom edges. */
	spacingVertical?: Space;
}

export function useSpacing({
	spacing,
	spacingBottom,
	spacingEnd,
	spacingHorizontal,
	spacingStart,
	spacingTop,
	spacingVertical,
}: SpacingProps): UtilityStyles {
	const cx = useStyles(styleSheet);
	const theme = useTheme();
	const map = useMemo(
		() => ({
			all: spacing,
			bottom: spacingBottom,
			end: spacingEnd,
			horizontal: spacingHorizontal,
			start: spacingStart,
			top: spacingTop,
			vertical: spacingVertical,
		}),
		[
			spacing,
			spacingBottom,
			spacingEnd,
			spacingHorizontal,
			spacingStart,
			spacingTop,
			spacingVertical,
		],
	);

	const className = useMemo(
		() =>
			cx(
				mapObject(map, (value) => (typeof value === 'number' ? 'inline' : value)),
				'element',
			),
		[cx, map],
	);

	const inlineStyle = useMemo(
		() =>
			mapObject(
				map,
				(value) => (typeof value === 'number' ? theme.unit(value) : undefined),
				(key) => `--spacing-${key}`,
			),
		[theme, map],
	);

	return {
		className,
		style: inlineStyle,
	};
}

export const styleSheet = style((css) => {
	const sizes: InlineSpacingSize[] = ['inline', ...SPACING_SIZES];

	function value(size: InlineSpacingSize, type: string) {
		return size === 'inline'
			? `var(--spacing-${type})`
			: css.var(`spacing-${size}` as 'spacing-df');
	}

	return {
		// Ordered from lowest to highest specificity
		'@variants': {
			...mapVariants('all', sizes, (size, type) => ({
				padding: value(size, type),
			})),

			...mapVariants('horizontal', sizes, (size, type) => ({
				paddingLeft: value(size, type),
				paddingRight: value(size, type),
			})),

			...mapVariants('vertical', sizes, (size, type) => ({
				paddingTop: value(size, type),
				paddingBottom: value(size, type),
			})),

			...mapVariants('top', sizes, (size, type) => ({
				paddingTop: value(size, type),
			})),

			...mapVariants('bottom', sizes, (size, type) => ({
				paddingBottom: value(size, type),
			})),

			...mapVariants('start', sizes, (size, type) => ({
				paddingLeft: value(size, type),
			})),

			...mapVariants('end', sizes, (size, type) => ({
				paddingRight: value(size, type),
			})),
		},
	};
});
