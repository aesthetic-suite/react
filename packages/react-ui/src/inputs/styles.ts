import { PaletteType, style } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const styleSheet = style((css) => ({
	width: '100%',
	display: 'block',
	color: css.var('palette-neutral-text'),
	backgroundColor: css.var('palette-neutral-color-00'),
	borderColor: css.var(`palette-neutral-bg-base`),
	borderStyle: 'solid',
	appearance: 'none',
	outline: 'none',
	margin: 0,
	padding: 0,
	zIndex: 0,

	':disabled': {
		cursor: 'default',
	},

	'::placeholder': {
		color: css.var('palette-muted-text'),
	},

	'@variants': {
		...mapVariants(
			'palette',
			['negative', 'positive', 'primary', 'warning'] as PaletteType[],
			(palette) => ({
				borderColor: css.var(`palette-${palette}-bg-base`),

				':focus': { borderColor: css.var(`palette-${palette}-bg-focused`) },
				':hover': { borderColor: css.var(`palette-${palette}-bg-hovered`) },

				':disabled': {
					borderColor: css.var(`palette-${palette}-bg-disabled`),
					color: css.var(`palette-${palette}-fg-disabled`),
				},
			}),
		),
	},
}));
