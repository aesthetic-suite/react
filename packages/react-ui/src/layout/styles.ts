import { styles } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';
import { BoxAlignContent, BoxAlignItems, BoxJustifyContent } from './types';

export const styleSheet = styles(() => ({
	box: {
		display: 'flex',
		margin: 0,
		padding: 0,
		listStyle: 'none',

		'@variants': {
			// Align content
			...mapVariants(
				'alignContent',
				[
					'baseline',
					'center',
					'end',
					'flex-end',
					'flex-start',
					'normal',
					'safe',
					'space-around',
					'space-between',
					'space-evenly',
					'start',
					'stretch',
					'unsafe',
				] as BoxAlignContent[],
				(value) => ({
					alignContent: value,
				}),
			),

			// Align items
			...mapVariants(
				'alignItems',
				[
					'baseline',
					'center',
					'end',
					'flex-end',
					'flex-start',
					'normal',
					'safe',
					'self-end',
					'self-start',
					'start',
					'stretch',
					'unsafe',
				] as BoxAlignItems[],
				(value) => ({
					alignItems: value,
				}),
			),

			// Align self
			...mapVariants(
				'alignSelf',
				[
					'auto',
					'baseline',
					'center',
					'flex-end',
					'flex-start',
					'normal',
					'safe',
					'self-end',
					'self-start',
					'stretch',
					'unsafe',
				],
				(value) => ({
					alignSelf: value,
				}),
			),

			// Flex direction
			'direction:column': { flexDirection: 'column' },
			'direction:column-reverse': { flexDirection: 'column-reverse' },
			'direction:row': { flexDirection: 'row' },
			'direction:row-reverse': { flexDirection: 'row-reverse' },

			// Justify content
			...mapVariants(
				'justifyContent',
				[
					'center',
					'end',
					'flex-end',
					'flex-start',
					'left',
					'normal',
					'right',
					'safe',
					'space-around',
					'space-between',
					'space-evenly',
					'start',
					'unsafe',
				] as BoxJustifyContent[],
				(value) => ({
					justifyContent: value,
				}),
			),

			// Flex wrap
			'wrap:false': { flexWrap: 'nowrap' },
			'wrap:true': { flexWrap: 'wrap' },
			'wrap:reverse': { flexWrap: 'wrap-reverse' },
		},
	},

	boxInline: {
		display: 'inline-flex',
	},

	boxVars: {
		// @ts-expect-error csstype not typed for variables
		flexGrow: 'var(--box-grow)',
		// @ts-expect-error csstype not typed for variables
		flexShrink: 'var(--box-shrink)',
		// @ts-expect-error csstype not typed for variables
		order: 'var(--box-order)',
	},

	inlineGap: {
		'@selectors': {
			'> *:not(:last-child)': {
				marginRight: 'var(--gap)',
			},
		},
	},

	stackGap: {
		'@selectors': {
			'> *:not(:last-child)': {
				marginBottom: 'var(--gap)',
			},
		},
	},
}));
