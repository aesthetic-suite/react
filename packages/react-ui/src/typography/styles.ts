import { PALETTE_TYPES, styles } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const styleSheet = styles((css) => {
	function weight(size: number | 'inherit') {
		return {
			fontWeight: size,
		};
	}

	return {
		typography: {
			color: css.var('palette-neutral-text'),
			margin: 0,
			padding: 0,

			'@variants': {
				// Alignment
				'align:start': {
					textAlign: 'start',
					'@fallbacks': {
						textAlign: 'left',
					},
				},
				'align:center': {
					textAlign: 'center',
				},
				'align:end': {
					textAlign: 'end',
					'@fallbacks': {
						textAlign: 'right',
					},
				},
				'align:justify': {
					textAlign: 'justify',
				},

				// Overflow
				'overflow:break': css.mixin('text-break'),
				'overflow:clip': css.mixin('text-truncate', {
					textOverflow: 'clip',
				}),
				'overflow:truncate': css.mixin('text-truncate'),
				'overflow:wrap': css.mixin('text-wrap'),

				// Palettes
				...mapVariants('palette', PALETTE_TYPES, (palette) => ({
					color: css.var(`palette-${palette}-text`),
				})),

				// Transforms
				'transform:capitalize': { textTransform: 'capitalize' },
				'transform:lowercase': { textTransform: 'lowercase' },
				'transform:uppercase': { textTransform: 'uppercase' },

				// Font weights
				'weight:inherit': weight('inherit'),
				'weight:thin': weight(100),
				'weight:light': weight(300),
				'weight:normal': weight(400),
				'weight:bold': weight(600),
				'weight:black': weight(800),
			},
		},

		monospace: {
			fontFamily: css.var('typography-font-monospace'),
		},

		link: {
			display: 'inline-block',
			backgroundColor: 'transparent',
			borderColor: 'transparent',
			borderWidth: 0,
			cursor: 'pointer',
			textDecoration: 'none',
			fontFamily: 'inherit',
			fontSize: 'inherit',
			lineHeight: 'inherit',

			'@variants': {
				// Palettes
				...mapVariants('palette', PALETTE_TYPES, (palette) => ({
					color: css.var(`palette-${palette}-text`),
					'@selectors': {
						':visited': { color: css.var(`palette-${palette}-color-60`) },
						':hover': { color: css.var(`palette-${palette}-color-80`) },
						':active': { color: css.var(`palette-${palette}-color-90`) },
					},
				})),
			},
		},
	};
});
