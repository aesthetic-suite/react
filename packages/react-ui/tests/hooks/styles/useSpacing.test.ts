import { renderHook } from '@testing-library/react-hooks';
import { useSpacing } from '../../../src/hooks/styles/useSpacing';
import { getRenderedStyles, withStyles, Wrapper } from '../../helpers';

describe('useSpacing()', () => {
	it('renders default', () => {
		const { result } = renderHook(() => useSpacing({}), { wrapper: Wrapper });

		expect(result.current).toEqual({ className: 'element', style: {} });
	});

	it(
		'renders default styles',
		withStyles(() => {
			renderHook(() => useSpacing({}), { wrapper: Wrapper });

			expect(getRenderedStyles('standard')).toMatchSnapshot();
		}),
	);

	it('can set all sides', () => {
		const { result } = renderHook(() => useSpacing({ spacing: 'df' }), { wrapper: Wrapper });

		expect(result.current).toEqual({ className: 'element variant:all:df', style: {} });
	});

	it('can set horizontal sides', () => {
		const { result } = renderHook(() => useSpacing({ spacingHorizontal: 'xs' }), {
			wrapper: Wrapper,
		});

		expect(result.current).toEqual({ className: 'element variant:horizontal:xs', style: {} });
	});

	it('can set vertical sides', () => {
		const { result } = renderHook(() => useSpacing({ spacingVertical: 'xl' }), {
			wrapper: Wrapper,
		});

		expect(result.current).toEqual({ className: 'element variant:vertical:xl', style: {} });
	});

	it('can set top side', () => {
		const { result } = renderHook(() => useSpacing({ spacingTop: 'sm' }), {
			wrapper: Wrapper,
		});

		expect(result.current).toEqual({ className: 'element variant:top:sm', style: {} });
	});

	it('can set right side', () => {
		const { result } = renderHook(() => useSpacing({ spacingEnd: 'df' }), {
			wrapper: Wrapper,
		});

		expect(result.current).toEqual({ className: 'element variant:end:df', style: {} });
	});

	it('can set bottom side', () => {
		const { result } = renderHook(() => useSpacing({ spacingBottom: 'lg' }), {
			wrapper: Wrapper,
		});

		expect(result.current).toEqual({ className: 'element variant:bottom:lg', style: {} });
	});

	it('can set left side', () => {
		const { result } = renderHook(() => useSpacing({ spacingStart: 'xl' }), {
			wrapper: Wrapper,
		});

		expect(result.current).toEqual({ className: 'element variant:start:xl', style: {} });
	});

	it('can set multiple sides with correct specificity', () => {
		const { result } = renderHook(
			() => useSpacing({ spacing: 'df', spacingVertical: 'sm', spacingTop: 'xs' }),
			{
				wrapper: Wrapper,
			},
		);

		expect(result.current).toEqual({
			className: 'element variant:all:df variant:vertical:sm variant:top:xs',
			style: {},
		});
	});

	it('can set sides using a number', () => {
		const { result } = renderHook(
			() =>
				useSpacing({
					spacing: 1,
					spacingHorizontal: 0.5,
					spacingVertical: 2,
					spacingTop: 3,
					spacingBottom: 5,
					spacingStart: 0,
					spacingEnd: -1,
				}),
			{
				wrapper: Wrapper,
			},
		);

		expect(result.current).toEqual({
			className:
				'element variant:all:inline variant:horizontal:inline variant:vertical:inline variant:top:inline variant:bottom:inline variant:start:inline variant:end:inline',
			style: {
				'--spacing-all': '1.25rem',
				'--spacing-bottom': '6.25rem',
				'--spacing-end': '-1.25rem',
				'--spacing-horizontal': '0.63rem',
				'--spacing-start': '0rem',
				'--spacing-top': '3.75rem',
				'--spacing-vertical': '2.50rem',
			},
		});
	});
});
