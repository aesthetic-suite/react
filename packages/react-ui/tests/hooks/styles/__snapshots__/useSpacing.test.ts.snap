// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSpacing() renders default styles 1`] = `".a {padding: var(--spacing-all);}.b {padding: var(--spacing-xs);}.c {padding: var(--spacing-sm);}.d {padding: var(--spacing-df);}.e {padding: var(--spacing-lg);}.f {padding: var(--spacing-xl);}.g {padding-left: var(--spacing-horizontal);}.h {padding-right: var(--spacing-horizontal);}.i {padding-left: var(--spacing-xs);}.j {padding-right: var(--spacing-xs);}.k {padding-left: var(--spacing-sm);}.l {padding-right: var(--spacing-sm);}.m {padding-left: var(--spacing-df);}.n {padding-right: var(--spacing-df);}.o {padding-left: var(--spacing-lg);}.p {padding-right: var(--spacing-lg);}.q {padding-left: var(--spacing-xl);}.r {padding-right: var(--spacing-xl);}.s {padding-top: var(--spacing-vertical);}.t {padding-bottom: var(--spacing-vertical);}.u {padding-top: var(--spacing-xs);}.v {padding-bottom: var(--spacing-xs);}.w {padding-top: var(--spacing-sm);}.x {padding-bottom: var(--spacing-sm);}.y {padding-top: var(--spacing-df);}.z {padding-bottom: var(--spacing-df);}.a1 {padding-top: var(--spacing-lg);}.b1 {padding-bottom: var(--spacing-lg);}.c1 {padding-top: var(--spacing-xl);}.d1 {padding-bottom: var(--spacing-xl);}.e1 {padding-top: var(--spacing-top);}.f1 {padding-top: var(--spacing-xs);}.g1 {padding-top: var(--spacing-sm);}.h1 {padding-top: var(--spacing-df);}.i1 {padding-top: var(--spacing-lg);}.j1 {padding-top: var(--spacing-xl);}.k1 {padding-bottom: var(--spacing-bottom);}.l1 {padding-bottom: var(--spacing-xs);}.m1 {padding-bottom: var(--spacing-sm);}.n1 {padding-bottom: var(--spacing-df);}.o1 {padding-bottom: var(--spacing-lg);}.p1 {padding-bottom: var(--spacing-xl);}.q1 {padding-left: var(--spacing-start);}.r1 {padding-left: var(--spacing-xs);}.s1 {padding-left: var(--spacing-sm);}.t1 {padding-left: var(--spacing-df);}.u1 {padding-left: var(--spacing-lg);}.v1 {padding-left: var(--spacing-xl);}.w1 {padding-right: var(--spacing-end);}.x1 {padding-right: var(--spacing-xs);}.y1 {padding-right: var(--spacing-sm);}.z1 {padding-right: var(--spacing-df);}.a2 {padding-right: var(--spacing-lg);}.b2 {padding-right: var(--spacing-xl);}"`;
